File: learn/stack-and-queue/stack-linked-list/stack_linked_list.go


type StackNode struct {
	data int
	next *StackNode
}

type StackLinkedList struct {
	top *StackNode
}

func (s *StackLinkedList) Push(value int) {
	newNode := &StackNode{data: value, next: s.top}
	s.top = newNode
}

func (s *StackLinkedList) Pop() int {
	if s.IsEmpty() {
		return -1
	}
	value := s.top.data
	s.top = s.top.next
	return value
}

func (s *StackLinkedList) Peek() int {
	if s.IsEmpty() {
		return -1
	}
	return s.top.data
}

func (s *StackLinkedList) IsEmpty() bool {
	return s.top == nil
}

File: learn/stack-and-queue/stack-linked-list/stack_linked_list.rs
pub struct StackNode {
    data: i32,
    next: Option<Box<StackNode>>,
}

pub struct StackLinkedList {
    top: Option<Box<StackNode>>,
}

impl StackLinkedList {
    pub fn new() -> Self {
        StackLinkedList { top: None }
    }

    pub fn push(&mut self, value: i32) {
        let new_node = Box::new(StackNode {
            data: value,
            next: self.top.take(),
        });
        self.top = Some(new_node);
    }

    pub fn pop(&mut self) -> Option<i32> {
        self.top.take().map(|node| {
            self.top = node.next;
            node.data
        })
    }

    pub fn peek(&self) -> Option<&i32> {
        self.top.as_ref().map(|node| &node.data)
    }

    pub fn is_empty(&self) -> bool {
        self.top.is_none()
    }
}
